src/emmy/numerical/ode.cljc

```diff
@@ -2,96 +2,171 @@
 
 (ns emmy.numerical.ode
   "ODE solvers for working with initial value problems."
-  (:require #?(:cljs ["odex" :as o])
-            #?(:cljs [emmy.util :as u])
-            #?(:cljs [emmy.value :as v])
+  (:require #?@(:cljs
+                [["odex" :as o]
+                 [emmy.util :as u]])
+            #?(:clj
+               [clojure.core.async :as a])
             [emmy.expression.compile :as c]
             [emmy.structure :as struct]
             [emmy.util.stopwatch :as us]
+            [emmy.value :as v]
             [taoensso.timbre :as log])
   #?(:clj
      (:import (org.apache.commons.math3.ode.nonstiff GraggBulirschStoerIntegrator)
               (org.apache.commons.math3.ode FirstOrderDifferentialEquations)
+              (org.apache.commons.math3.ode ExpandableStatefulODE)
               (org.apache.commons.math3.ode.sampling StepHandler))))
 
-#?(:cljs
-   (def ^:private near? (v/within 1e-8)))
 
-#?(:clj
-   (defn- round-up
-     "Returns `n` rounded up to the nearest multiple of `step-size`. the returned
-  value will always equal `0`, mod `step-size`"
-     [n step-size]
-     (let [offset (mod n step-size)]
-       (if (pos? offset)
-         (-> (- n offset)
-             (+ step-size))
-         n))))
 
-#?(:clj
-   (defn step-handler
-     "Generates a StepHandler instance that can be attached to an integrator.
+(def ^:private near? (v/within 1e-8))
+(def ^:private default-epsilon 1e-8)
+
+(defn stream-integrator
+  "Produces a function, monotonic in its single numeric argument, 
+   that represents the integral of the function f' given the initial 
+   data $y_0 = f(x_0)$ and an options dictionary (presently containing
+   the tolerance for error $\\epsilon$, but eventually also selecting
+   from a menu of integration techniques). 
+   
+   This is done by creating an adaptive step-size ODE solver, and 
+   advancing its steps as needed to supply function values. (This 
+   architecture accounts for why the arguments to f must be presented
+   in order). Old solution segments are discarded. The goal of this 
+   approach is to avoid the requirement of supplying an upper limit
+   to the integration. At the cost of requiring monotonic arguments
+   to f, the integrated function can essentially be used forever
+   without accumulating unbounded state.
+   
+   The returned function may be called with no arguments to shut down
+   the integration, allowing for the final reclamation of its resources.
+
+   When the ODE solver is provided by Java, it may be necessary to 
+   use an auxiliary thread to enable this style of flow control.  If
+   JavaScript, we expect the solver to provide a generator of solution
+   segments."
+  [f' x0 y0 {:keys [epsilon] :or {epsilon default-epsilon}}]
```

Comment:
are we letting `y0` be a structure here? If so we can use `emmy.structure/dimension`.

Reply:
No. `stream-integrator` by design offers a `double-array -> double-array` interface, leaving the SICM layer to adapt state-derivatives in this form.